### PR TITLE
[Segmenter] test the groups the same order as they are merged

### PR DESCRIPTION
**Pull request by shmsong**
This PR is a work around for #1947. It makes sure that the input and output ordering of the merged groups stay consistent in different phases of segment pass so that `canSchedule` returns consistent results.

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
I can confirm that with these changes both code snippets from the issue description and https://github.com/csarofeen/pytorch/issues/1947#issuecomment-1232890228 now work.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:2860]: This change is counter-intuitive.

The old API only tries to merge all candidate connected by `candidate_it->edge`. Shouldn't we plumb the connection check to `getMergeCandidates` instead of assert on the connectivity inside `codeGenSupportedMerge`?

**Comment by csarofeen** [comment]
Which phases were inconsistent? Is this if we identify a segment is good, then we make the segment, the output ordering isn't the same because `supportedMerge` and `makeFusion` can have different ordering of outputs?

**Review by csarofeen — APPROVED** [review]
I think this PR seems fine, if I understand correctly you're just making sure processing an edge we don't merge groups as `(consumer, producer)` but always merge as `(producer, consumer)`.

**Comment by csarofeen** [comment]
This seems fine as a WAR, though we definitely should figure out what's going on in https://github.com/csarofeen/pytorch/issues/1947.

File: torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp:2860]: `getMergeCandidates` already guarantees that they are connected by construction, since looking at immediate neighbors is exactly what it does. 

the connectivity on `codeGenSupportedMerge` I added isn't strictly necessary but just a sanity check that `edge` -> `group1, group2` is semantically equivalent to what we've been doing. 

I'd expect removing that assert in future PR's when we try to test more candidate patterns, which hasn't been prioritized.

**Comment by shmsong** [comment]
Merging as is for now to unblock.

**Pull request merged**
